### PR TITLE
RValue, prepared constructor

### PR DESCRIPTION
**Pull request by Lorak-mmk**
Added overloads to execute accepting ravlue rerefences.
Added constructor to prepared query, for easier usage as a class/struct member

### CODE REVIEW

File: include/batch_query.hh [hunk 1]
```diff
@@ -1,9 +1,13 @@
...
-#include "cassandra.h"
 
-#include "statement.hh"
+#include <string>
+
+#include "cassandra.h"
 
 namespace scmd {
+
+class statement;
```
Comment (jbhayven) [inline include/batch_query.hh:9]: Why this way?

File: include/session.hh [hunk 2]
```diff
@@ -27,63 +27,104 @@ public:
...
 
     template<typename... Args,
         typename = typename std::enable_if<0 != sizeof...(Args)>::type>
     future execute_async(const std::string &query, Args... args) {
         return execute_async(scmd::statement(query, sizeof...(args)).bind(args...));
     }
+    template<typename... Args,
+        typename = typename std::enable_if<0 != sizeof...(Args)>::type>
+    future execute_async(const std::string &&query, Args... args) {
+        return execute_async(scmd::statement(query, sizeof...(args)).bind(args...));
```
Comment (jbhayven) [inline include/session.hh:55]: I think the query should be moved here further to the statement constructor, if possible.

File: include/session.hh [hunk 3]
```diff
@@ -27,63 +27,104 @@ public:
...
     future execute_async(statement &statement, Args... args) {
         return execute_async(statement.bind(args...));
     }
-
     template<typename... Args,
         typename = typename std::enable_if<0 != sizeof...(Args)>::type>
-    future execute_async(const prepared_query& query, Args... args) {
-        auto statement = query.get_statement();
+    future execute_async(const statement &&statement, Args... args) {
         return execute_async(statement.bind(args...));
```
Comment (jbhayven) [inline include/session.hh:67]: Not sure of how moving works here (what does 'bind' return? a value? or a reference? if it's a reference, is this what we really intend to do? well, perhaps yes, but still, careful with this).

File: include/session.hh [hunk 4]
```diff
@@ -27,63 +27,104 @@ public:
...
 
     template<typename... Args,
             typename = typename std::enable_if<0 != sizeof...(Args)>::type>
     query_result execute(const std::string &query, Args... args) {
         return execute_async(scmd::statement(query, sizeof...(args)).bind(args...)).get_result();
     }
+    template<typename... Args,
+        typename = typename std::enable_if<0 != sizeof...(Args)>::type>
+    query_result execute(const std::string &&query, Args... args) {
+        return execute_async(scmd::statement(query, sizeof...(args)).bind(args...)).get_result();
```
Comment (jbhayven) [inline include/session.hh:103]: Move the query.

File: src/session.cc [hunk 5]
```diff
@@ -42,43 +42,75 @@ session::~session() {
...
 }
+future session::execute_async(const statement &&statement) {
+    return future(cass_session_execute(_session, statement.get_statement()));
+}
 
 future session::execute_async(const std::string &query) {
     return execute_async(statement(query));
 }
+future session::execute_async(const std::string &&query) {
+    return execute_async(statement(query));
```
Comment (jbhayven) [inline src/session.cc:66]: Move the query further to the statement.

File: src/session.cc [hunk 6]
```diff
@@ -42,43 +42,75 @@ session::~session() {
...
+future session::execute_async(const batch_query &&query) {
+    return future(cass_session_execute_batch(_session, query.get_query()));
+}
 
 future session::execute_async(const prepared_query &query) {
     auto statement = query.get_statement();
     return execute_async(statement);
 }
+future session::execute_async(const prepared_query &&query) {
+    return execute_async(query.get_statement());
```
Comment (jbhayven) [inline src/session.cc:81]: Move the statement? If possible. Again, this is one of those cases where I don't know what to do.

File: src/session.cc [hunk 7]
```diff
@@ -42,43 +42,75 @@ session::~session() {
...
+}
 
 
 query_result session::execute(const statement &statement) {
     return execute_async(statement).get_result();
 }
+query_result session::execute(const statement &&statement) {
+    return execute_async(statement).get_result();
```
Comment (jbhayven) [inline src/session.cc:89]: Move the statement further.

File: src/session.cc [hunk 8]
```diff
@@ -42,43 +42,75 @@ session::~session() {
...
+}
+
 
 query_result session::execute(const std::string &query) {
     return execute_async(query).get_result();
 }
+query_result session::execute(const std::string &&query) {
+    return execute_async(query).get_result();
```
Comment (jbhayven) [inline src/session.cc:97]: Move the query further.

File: src/session.cc [hunk 9]
```diff
@@ -42,43 +42,75 @@ session::~session() {
...
 
 query_result session::execute(const batch_query &query) {
     return execute_async(query).get_result();
 }
-
-prepared_query session::prepare(const std::string& query) {
-    scmd::future future(cass_session_prepare(_session, query.c_str()));
-    return future.get_prepared();
+query_result session::execute(const batch_query &&query) {
+    return execute_async(query).get_result();
```
Comment (jbhayven) [inline src/session.cc:104]: Move.

File: src/session.cc [hunk 10]
```diff
@@ -42,43 +42,75 @@ session::~session() {
...
 }
+
 query_result session::execute(const prepared_query &query) {
     auto statement = query.get_statement();
     return execute(statement);
 }
+query_result session::execute(const prepared_query &&query) {
+    auto statement = query.get_statement();
+    return execute(statement);
```
Comment (jbhayven) [inline src/session.cc:113]: Why not execute(query.get_statement())? (Or even execute(std::move(query.get_statement)), if possible).

**Review by jbhayven — COMMENTED** [review]
An idea for testing the move semantics: subclass a structure, add a static counter that is increased by the standard and copy constructor, and control how many times it is done when the structure is passed to be executed. 

If an execution of such a test_statement does not modify the counter then, I suppose, the move semantics is done well. (I may be wrong, though).

**Review by jbhayven — APPROVED** [review]
Looks more or less okay now. However, it would be great if someone better verse in C++ could take a look as well.

**Review by Rhantolq — APPROVED** [review]

**Pull request merged**